crates/base-db/src/input.rs

```diff
@@ -572,25 +686,41 @@ impl CrateGraph {
     /// Note that for deduplication to fully work, `self`'s crate dependencies must be sorted by crate id.
     /// If the crate dependencies were sorted, the resulting graph from this `extend` call will also have the crate dependencies sorted.
     pub fn extend(&mut self, mut other: CrateGraph, proc_macros: &mut ProcMacroPaths) {
+        enum ExtendStrategy {
+            Dedup(CrateId),
+            Replace(CrateId),
+        }
+
         let topo = other.crates_in_topological_order();
         let mut id_map: FxHashMap<CrateId, CrateId> = FxHashMap::default();
-
         for topo in topo {
             let crate_data = &mut other.arena[topo];
-            crate_data.dependencies.iter_mut().for_each(|dep| dep.crate_id = id_map[&dep.crate_id]);
-            crate_data.dependencies.sort_by_key(|dep| dep.crate_id);
 
-            let res = self.arena.iter().find_map(
-                |(id, data)| {
-                    if data == crate_data {
-                        Some(id)
-                    } else {
-                        None
+            crate_data.dependencies.iter_mut().for_each(|dep| {
+                dep.crate_id = id_map[&dep.crate_id];
+            });
+            crate_data.dependencies.sort_by_key(|dep| dep.crate_id);
+            let res = self.arena.iter().find_map(|(id, data)| {
+                if data.almost_eq(crate_data) {
+                    if data.origin.is_lib() && crate_data.origin.is_local() {
+                        // See #15656 for a relevant example.
+                        return Some(ExtendStrategy::Replace(id));
```

Comment:
`Replace` is a strong word for us effectively wanting to deduplicate with an `origin` overwrite/preference. So instead of creating a new crate and removing the previous one, let's just update the duplicates origin below in the match.